apps/documents/models.py

```diff
@@ -113,20 +110,92 @@ def version_details(self) -> VersionDetails:
         self.save(update_fields=["version_number"])
 
         new_version = super().create_new_version(save=False)
         new_version.version_number = version_number
+        vector_store_present = bool(new_version.openai_vector_store_id)
+        new_version.openai_vector_store_id = ""
         new_version.save()
 
         file_versions = []
         for file in self.files.iterator(chunk_size=15):
-            file_versions.append(file.create_new_version())
+            file_version = file.create_new_version(save=False)
+            file_version.external_id = ""
+            file_version.external_source = ""
+            file_version.save()
+            file_versions.append(file_version)
 
         new_version.files.add(*file_versions)
+
+        if self.is_index and vector_store_present:
+            # Create vector store at llm service
+            # Optimization suggestion: Only when the file set changed, should we create a new vector store at the
+            # provider
+            manager = new_version.llm_provider.get_index_manager()
+            version_name = f"{new_version.index_name} v{new_version.version_number}"
+            new_version.openai_vector_store_id = manager.create_vector_store(name=version_name)
+            new_version.save(update_fields=["openai_vector_store_id"])
+
+            # Upload files to vector store
+            index_collection_files(new_version.id, all_files=True)
+
         return new_version
 
     def get_absolute_url(self):
         return reverse("documents:single_collection_home", args=[self.team.slug, self.id])
+
+    def get_related_nodes_queryset(self) -> models.QuerySet:
+        index_references = get_related_pipelines_queryset(self, "collection_index_id").distinct()
+        collection_references = get_related_pipelines_queryset(self, "collection_id").distinct()
+        return index_references | collection_references
+
+    def get_related_experiments_queryset(self) -> models.QuerySet:
+        """
+        Get all experiments that reference this collection through a pipeline
+        """
+        # TODO: Update assistant archive code to use get_related_pipeline_experiments_queryset
+        index_references = get_related_pipeline_experiments_queryset(self, "collection_index_id").filter(
+            is_default_version=True
+        )
+        collection_references = get_related_pipeline_experiments_queryset(self, "collection_id").filter(
+            is_default_version=True
+        )
+        return index_references | collection_references
+
+    @transaction.atomic()
+    def archive(self):
+        """
+        Archive the collection with its files and remove the index and the files at the remote service, if it has one
+        """
+        if self.get_related_nodes_queryset().exists():
+            return False
+
+        if self.is_working_version:
+            for version in self.versions.all():
+                if version.get_related_experiments_queryset().exists():
```

Comment:
Can't this be done without querying in a loop? We could fetch the version IDs and then run the query with that:

```python
version_ids = self.versions.values_list("id", flat=True)
has_references = Node.objects.filter(params__collection_id__in=version_ids).exists()
```

Reply:
Thoughts on https://github.com/dimagi/open-chat-studio/pull/1587/commits/be84bd635bf144cb3b11c13a4d0915b52ce6d77d?